File: java/data-structure/Stack/src/main/java/ru/reimu/stack/ArrayStack.java
package ru.reimu.stack;

import ru.reimu.array.Array;

/**
 * @Author: Tomonori
 * @Date: 2020/2/14 17:59
 * @Title:
 * @Desc: ↓ ↓ ↓ ↓ ↓
 * -----
 */
public class ArrayStack<T> implements Stack<T> {

    Array<T> array;

    public ArrayStack() {
        this.array = new Array<>();
    }

    public ArrayStack(int capacity) {
        this.array = new Array<>(capacity);
    }

    public int getCapacity() {
        return array.getCap();
    }

    @Override
    public int getSize() {
        return array.getSize();
    }

    @Override
    public boolean isEmpty() {
        return array.isEmpty();
    }

    @Override
    public void push(T t) {
        array.addLast(t);
    }

    @Override
    public T pop() {
        return array.removeLast();
    }

    @Override
    public T peek() {
        return array.getLast();
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("Stack: ");
        sb.append('[');

        for (int i = 0; i < array.getSize(); i++) {
            sb.append(array.get(i));
            if (i != array.getSize() - 1) {
                sb.append(", ");
            }
        }
        sb.append("] top");

        return sb.toString();
    }
}

File: golang/data-structure/stack/ArrayStack.go
package stack

import (
	array2 "all/array/array"
	"bytes"
	"fmt"
)

func NewArrayStack() *ArrayStack {
	object := array2.NewArray()
	return &ArrayStack{array: object}
}

func NewArrayStackWithCap(cap int) *ArrayStack {
	object := array2.NewArrayWithCap(cap)
	return &ArrayStack{array: object}
}

type ArrayStack struct {
	array *array2.Array
}

func (a *ArrayStack) GetCap() int {
	return a.array.GetCap()
}

func (a *ArrayStack) GetSize() int {
	return a.array.GetSize()
}

func (a *ArrayStack) IsEmpty() bool {
	return a.array.IsEmpty()
}

func (a *ArrayStack) Push(t interface{}) error {
	return a.array.AddLast(t)
}

func (a *ArrayStack) Pop() (interface{}, error) {
	return a.array.RemoveLast()
}

func (a *ArrayStack) Peek() (interface{}, error) {
	return a.array.GetLast()
}

func (a *ArrayStack) String() string {
	buffer := bytes.Buffer{}
	buffer.WriteString("stack: ")
	buffer.WriteByte('[')

	for i := 0; i < a.GetSize(); i++ {
		if i < a.GetSize() {
			item, _ := a.array.Get(i)
			buffer.WriteString(fmt.Sprintf("%v", item))
			if i != a.GetSize()-1 {
				buffer.WriteString(", ")
			}
		}
	}

	buffer.WriteString("] on top")

	return buffer.String()
}
